rcl/src/rcl/wait.c

```diff
@@ -0,0 +1,448 @@
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    allocator.deallocate, "deallocate not set", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    allocator.reallocate, "reallocate not set", return RCL_RET_INVALID_ARGUMENT);
+  // Allocate space for the implementation struct.
+  wait_set->impl = (rcl_wait_set_impl_t *)allocator.allocate(
+    sizeof(rcl_wait_set_impl_t), allocator.state);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    wait_set->impl, "allocating memory failed", return RCL_RET_BAD_ALLOC);
+  wait_set->impl->rmw_subscriptions.subscribers = NULL;
+  wait_set->impl->rmw_subscriptions.subscriber_count = 0;
+  wait_set->impl->rmw_guard_conditions.guard_conditions = NULL;
+  wait_set->impl->rmw_guard_conditions.guard_condition_count = 0;
+  // Initialize subscription space.
+  rcl_ret_t ret;
+  if ((ret = rcl_wait_set_resize_subscriptions(wait_set, number_of_subscriptions)) != RCL_RET_OK) {
+    fail_ret = ret;
+    goto fail;
+  }
+  if ((ret = rcl_wait_set_clear_subscriptions(wait_set)) != RCL_RET_OK) {
+    fail_ret = ret;
+    goto fail;
+  }
+  // Initialize guard condition space.
+  ret = rcl_wait_set_resize_guard_conditions(wait_set, number_of_guard_conditions);
+  if (ret != RCL_RET_OK) {
+    fail_ret = ret;
+    goto fail;
+  }
+  if ((ret = rcl_wait_set_clear_guard_conditions(wait_set)) != RCL_RET_OK) {
+    fail_ret = ret;
+    goto fail;
+  }
+  // Initialize timer space.
+  ret = rcl_wait_set_resize_timers(wait_set, number_of_timers);
+  if (ret != RCL_RET_OK) {
+    fail_ret = ret;
+    goto fail;
+  }
+  if ((ret = rcl_wait_set_clear_timers(wait_set)) != RCL_RET_OK) {
+    fail_ret = ret;
+    goto fail;
+  }
+  // Set allocator.
+  wait_set->impl->allocator = allocator;
+  return RCL_RET_OK;
+fail:
+  __wait_set_clean_up(wait_set, allocator);
+  return fail_ret;
+}
+
+rcl_ret_t
+rcl_wait_set_fini(rcl_wait_set_t * wait_set)
+{
+  rcl_ret_t result = RCL_RET_OK;
+  RCL_CHECK_ARGUMENT_FOR_NULL(wait_set, RCL_RET_INVALID_ARGUMENT);
+  if (__wait_set_is_valid(wait_set)) {
+    __wait_set_clean_up(wait_set, wait_set->impl->allocator);
+  }
+  return result;
+}
+
+rcl_ret_t
+rcl_wait_set_get_allocator(const rcl_wait_set_t * wait_set, rcl_allocator_t * allocator)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(wait_set, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(allocator, RCL_RET_INVALID_ARGUMENT);
+  if (!__wait_set_is_valid(wait_set)) {
+    RCL_SET_ERROR_MSG("wait set is invalid");
+    return RCL_RET_WAIT_SET_INVALID;
+  }
+  *allocator = wait_set->impl->allocator;
+  return RCL_RET_OK;
+}
+
+#define SET_ADD(Type) \
+  RCL_CHECK_ARGUMENT_FOR_NULL(wait_set, RCL_RET_INVALID_ARGUMENT); \
+  RCL_CHECK_ARGUMENT_FOR_NULL(Type, RCL_RET_INVALID_ARGUMENT); \
+  if (!__wait_set_is_valid(wait_set)) { \
+    RCL_SET_ERROR_MSG("wait set is invalid"); \
+    return RCL_RET_WAIT_SET_INVALID; \
+  } \
+  if (!(wait_set->impl->Type ## _index < wait_set->size_of_ ## Type ## s)) { \
```

Comment:
http://stackoverflow.com/questions/22975073/what-does-double-hash-do-in-a-preprocessor-directive